examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3785 @@
+        Amount::zero(),
+    )?;
+
+    // Log the newly minted tokens.
+    let parameter_bytes = to_bytes(&Cis2Event::Mint(MintEvent {
+        token_id: TOKEN_ID_WCCD,
+        amount:   ContractTokenAmount::from(amount.micro_ccd),
+        owner:    sender,
+    }));
+
+    host.invoke_contract_raw(
+        &proxy_address,
+        Parameter(&parameter_bytes),
+        EntrypointName::new_unchecked("logEvent"),
+        Amount::zero(),
+    )?;
+
+    // Only logs a transfer event if the receiver is not the sender.
+    // Only executes the `OnReceivingCis2` hook if the receiver is not the sender
+    // and the receiver is a contract.
+    if sender != receive_address {
+        // Log the transfer event.
+        let parameter_bytes = to_bytes(&Cis2Event::Transfer(TransferEvent {
+            token_id: TOKEN_ID_WCCD,
+            amount:   ContractTokenAmount::from(amount.micro_ccd),
+            from:     sender,
+            to:       receive_address,
+        }));
+
+        host.invoke_contract_raw(
+            &proxy_address,
+            Parameter(&parameter_bytes),
+            EntrypointName::new_unchecked("logEvent"),
+            Amount::zero(),
+        )?;
+
+        // If the receiver is a contract: invoke the receive hook function.
+        if let Receiver::Contract(address, function) = input.params.to {
+            let parameter = OnReceivingCis2Params {
+                token_id: TOKEN_ID_WCCD,
+                amount:   ContractTokenAmount::from(amount.micro_ccd),
+                from:     sender,
+                data:     input.params.data,
+            };
+            host.invoke_contract(
+                &address,
+                &parameter,
+                function.as_entrypoint_name(),
+                Amount::zero(),
+            )
+            .unwrap_abort();
+        }
+    }
+    Ok(())
+}
+
+/// Unwrap an amount of wCCD tokens into CCD
+#[receive(contract = "CIS2-wCCD", name = "unwrap", error = "ContractError", mutable)]
+fn contract_unwrap<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<()> {
+    let (proxy_address, state_address) = get_protocol_addresses_from_implementation(host)?;
+
+    // Can be only called through the fallback function on the proxy.
+    only_proxy(proxy_address, ctx.sender())?;
+
+    // Check that contract is not paused.
+    when_not_paused(&state_address, host)?;
+
+    let input: UnwrapParamsWithSender = ctx.parameter_cursor().get()?;
+
+    if input.params.amount == 0u64.into() {
+        return Ok(());
+    }
+
+    // Get the sender who invoked this contract function.
+    let sender = input.sender;
+
+    ensure!(
+        sender == input.params.owner
+            || host
+                .state()
+                .is_operator(&state_address, &sender, &input.params.owner, host)
+                .map_or(false, |is_operator| is_operator),
```

Comment:
If the state contract fails to respond, this will result in `Unauthorized`. It would be better to return the error.